ferris_elf/picoseconds.py

```diff
@@ -0,0 +1,62 @@
+from dataclasses import dataclass
+from typing import Self
+
+
+def format_picos(ts: float | int) -> str:
+    timestamp = float(ts)
+
+    base = "ps"
+    scalar: list[tuple[str, int]] = [
+        ("ns", 1000),
+        ("µs", 1000),
+        ("ms", 1000),
+        ("s", 1000),
+        ("m", 60),
+        ("h", 60),
+    ]
+
+    for name, offset in scalar:
+        if timestamp > offset:
+            timestamp /= offset
+            base = name
+        else:
+            break
+
+    # remove any trailing stuff
+    timestamp = round(timestamp, 2)
+
+    # fun awesome fact, int decays to float in typeck directly,
+    # but is_integer is only implemented on float so if we didn't
+    # explicitly call timestamp = float(ts), this could break
+    # and typeck would be ok with it :D
+    if timestamp.is_integer():
+        return f"{timestamp:.0f}{base}"
+    else:
+        return f"{timestamp:.2f}{base}"
+
+
+@dataclass
+class Picoseconds:
+    __slots__ = "_value"
+
+    _value: int
```

Comment:
why are we defining `__slots__` manually? we can just do `@dataclass(slots=True)`

Reply:
also we can do `frozen=True` in the dataclass constructor to make it `__hash__`, that will be better than just a privacy marker